### PR TITLE
Add dice loss

### PR DESCRIPTION
**Pull request by wanghaoshuang**

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
         })
     return pool_out
+
+
+def dice_loss(input, label, num_classes, reduce_dim=[1, 2, 3], epsilon=0.00001):
+    """
+    **Dice loss Layer**
+    Dice loss for comparing the similarity of two batch of data,
+    usually be used for binary image segmentation i.e. labels are binary.
+    The value between 0 to 1, 1 means totally match.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: > The value between 0 to 1

是什么值呢？ 语法也不对。

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+
+    Args:
+        input (Variable): The predictions with shape [batch_size, ... , num_classes].
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: The shape `[batch_size, ... , num_classes]` is not clear.  Is the rank large than 2 (>=2D)? The first dimension is batch size, the last dimension is class number?

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+        label (Variable): The groud truth with shape [batch_szie, ... , 1].
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: The rank of `label` is the same with `input`?

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+        num_classes (integer): The number of target classes.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: It seems the `num_classes` can be got from `input`.

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+        reduce_dim (list<int>): The dimensions to be reduced. It should be set to input.shape[1:].
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: >  It should be set to input.shape[1:].

If so,  the `reduce_dim`  can be set Inside this function? there is no need to be an argument?

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+                                Default: [1,2,3]
+        epsilon (float): The epsilon will be added to the numerator and denominator.
+                         If both input and label are empty, it makes sure dice is 1.
+                         Default: 0.00001
+
+    Returns:
+        dice_loss (Variable): The dice loss.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Give the shape for the returned value.

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+    """
+    helper = LayerHelper('dice_loss', **locals())
+    out = helper.create_tmp_variable(dtype=input.dtype)
+    label = one_hot(label, depth=num_classes)
+    inse = reduce_sum(input * label, dim=reduce_dim)
+
+    dice_denominator = reduce_sum(
+        input, dim=reduce_dim) + reduce_sum(
+            label, dim=reduce_dim)
+    dice_score = (inse * 2 + epsilon) / (dice_denominator + epsilon)
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: It seems the returned loss should be 1 - dice_score from https://github.com/pytorch/pytorch/issues/1249 .

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
...
+    return reduce_mean(dice_score)
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: It seems the returned loss should be 1 - dice_score from https://github.com/pytorch/pytorch/issues/1249 .

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (ZumingHuang) [reply python/paddle/fluid/layers/nn.py]: 应该是loss的值介于[0, 1]之间

Comment (zchengquan) [reply python/paddle/fluid/layers/nn.py]: loss值为0，才代表完全match, 公式化表达是 dice-loss = 1 - 2 * intersection_area / total_area = (total_area - intersection_area) - intersection_area = (union_area - intersection_area) / total_area

Comment (Yipeng-Sun) [reply python/paddle/fluid/layers/nn.py]: 补充增加第三个等式的分母 dice-loss = 1 - 2 * intersection_area / total_area =（ (total_area - intersection_area) - intersection_area) / total_area = (union_area - intersection_area) / total_area

File: python/paddle/fluid/layers/nn.py (hunk 8 above)
Comment (ZumingHuang) [reply python/paddle/fluid/layers/nn.py]: It may not be necessary to add epsilon to the numerator

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -3815,3 +3816,39 @@ def roi_pool(input, rois, pooled_height=1, pooled_width=1, spatial_scale=1.0):
             "spatial_scale": spatial_scale
         })
     return pool_out
+
+
+def dice_loss(input, label, num_classes, reduce_dim=[1, 2, 3], epsilon=0.00001):
+    """
+    **Dice loss Layer**
+    Dice loss for comparing the similarity of two batch of data,
+    usually be used for binary image segmentation i.e. labels are binary.
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed.Thx.

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

File: python/paddle/fluid/layers/nn.py (hunk 6 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

File: python/paddle/fluid/layers/nn.py (hunk 8 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Fixed. Thx.

Comment (KokeCacao) [reply python/paddle/fluid/layers/nn.py]: > It may not be necessary to add epsilon to the numerator

Should be necessary since prediction [0,0,0,0] and target [0,0,0,0] expect a perfect score.
